### PR TITLE
Simulation framework improvements

### PR DESCRIPTION
**Pull request by lmars**
This includes a number of simulation improvements and fixes including:

* Fix stopping exec nodes
* Fix using the Docker adapter
* Add live node stop events (/cc @holisticode)
* Don't log errors for duplicate events
* Prevent a race when shutting down the network (between disconnecting nodes, stopping them but them reconnecting to each other)

### CODE REVIEW

File: p2p/simulations/network.go [hunk 1]
```diff
@@ -528,6 +540,10 @@ func (self *Network) GetNode(id discover.NodeID) *Node {
 func (self *Network) GetNodeByName(name string) *Node {
 	self.lock.Lock()
 	defer self.lock.Unlock()
+	return self.getNodeByName(name)
+}
+
+func (self *Network) getNodeByName(name string) *Node {
 	for _, node := range self.Nodes {
```
Comment (zelig) [inline p2p/simulations/network.go:547]: map?

File: p2p/simulations/network.go [hunk 2]
```diff
@@ -589,14 +605,6 @@ func (self *Network) getConn(oneID, otherID discover.NodeID) *Conn {
 }
 
 func (self *Network) Shutdown() {
-	// disconnect all nodes
```
Comment (zelig) [inline p2p/simulations/network.go:592]: wonder why this was here actually

**Comment by zelig** [comment]
Great stuff @lmars 
why execadapter no longer embeded in docker?
was simply the spurious call to disconnects in `network.Conns` unnecassary? and in fact harmful introducing deadlock stuff?

File: p2p/simulations/network.go (hunk 2 above)
Comment (lmars) [reply p2p/simulations/network.go:592]: This is left over from when stopping sim nodes didn't actually drop connections. Sim nodes now drop connections when stopped as they use a p2p server, so we don't need this.

**Comment by lmars** [comment]
> why execadapter no longer embeded in docker?

This change is the opposite, I now embed `ExecAdapter` in `DockerAdapter` so that it also tracks nodes.

**Pull request merged**